File: problems/data_structures/hashtable/lc_3138.py
from collections import Counter

"""
1 <= s.length <= 10^5
"""


class Solution:
    def minAnagramLength(self, s: str) -> int:
        n = len(s)

        def check(m):
            hashtbl = Counter(s[:m])
            for i in range(m, n, m):
                tmp_tbl = {}
                for j in range(i, i + m):
                    if s[j] not in tmp_tbl:
                        tmp_tbl[s[j]] = 1
                    else:
                        tmp_tbl[s[j]] += 1
                    if s[j] not in hashtbl or tmp_tbl[s[j]] > hashtbl[s[j]]:
                        return False
                if len(tmp_tbl) != len(hashtbl):
                    return False
            return True

        for i in range(1, n):
            if n % i != 0:
                continue
            if check(i):
                return i
        return n


s = "abba"
assert Solution().minAnagramLength(s) == 2

s = "cdef"
assert Solution().minAnagramLength(s) == 4

File: problems/data_structures/hashtable/lc_3138.cpp
#include <string>
#include <unordered_map>

using std::string;
using std::unordered_map;

class Solution
{
public:
  int minAnagramLength(const string &s)
  {
    int n      = s.length();
    auto check = [&](int m) {
      unordered_map<char, int> basetbl;
      for (auto c : s.substr(0, m))
        ++basetbl[c];
      for (int i = m; i < n; i += m) {
        unordered_map<char, int> tmptbl;
        for (int j = i; j < i + m; ++j) {
          tmptbl[s[j]] += 1;
          if (basetbl.count(s[j]) == 0 || basetbl[s[j]] < tmptbl[s[j]]) {
            // basetbl.count(s[j]) < tmptbl[s[j]] is false
            // basetbl[s[j]] < tmptbl[s[j]] is also false
            // why?
            return false;
          }
        }
        if (basetbl.size() != tmptbl.size())
          return false;
      }
      return true;
    };

    for (int i = 1; i < n; ++i) {
      if (n % i != 0) continue;
      if (check(i)) return i;
    }
    return n;
  }
};